File: android/sdk/src/main/java/com/viant/forgeandroid/ui/TranscriptEnvelope.kt
package com.viant.forgeandroid.ui

import com.viant.forgeandroid.runtime.JsonUtil
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

@Serializable
data class TranscriptForgeDataBlock(
    val version: Int? = 1,
    val id: String? = null,
    val format: String? = null,
    val mode: String? = null,
    val data: JsonElement? = null
)

@Serializable
data class TranscriptForgeUiPayload(
    val version: Int? = 1,
    val title: String? = null,
    val subtitle: String? = null,
    val blocks: List<JsonObject> = emptyList()
)

data class TranscriptForgeDataStore(val id: String, val rows: Any?)

/**
 * Platform-neutral transcript parts emitted by a conversation SDK. Forge owns
 * their data/UI pairing so native hosts do not re-parse completed messages.
 */
data class TranscriptCanonicalData(
    val version: Int? = null,
    val scope: String? = null,
    val reportRef: String? = null,
    val sequence: Int? = null,
    val id: String,
    val format: String? = null,
    val mode: String? = null,
    val payload: JsonElement? = null
)

data class TranscriptCanonicalPart(
    val kind: String,
    val text: String? = null,
    val source: String? = null,
    val payload: JsonElement? = null,
    val data: TranscriptCanonicalData? = null
)

data class TranscriptCanonicalReport(
    val scope: String,
    val id: String,
    val grammar: String,
    val status: String,
    val sequence: Int? = null,
    val resetVersion: Int = 0,
    val source: JsonElement,
    val dataSources: Map<String, TranscriptCanonicalData> = emptyMap()
)

sealed interface TranscriptEnvelopePart {
    data class Markdown(val text: String) : TranscriptEnvelopePart
    data class ForgeUi(val payload: TranscriptForgeUiPayload, val dataStore: Map<String, TranscriptForgeDataStore>) : TranscriptEnvelopePart
}

/** Generic Forge transcript decoding. Hosts own placement, not this envelope protocol. */
object TranscriptEnvelope {
    private val json = Json { ignoreUnknownKeys = true }

    /**
     * Removes progressive report transport fences from text that accompanies
     * an already-assembled canonical report. Ordinary markdown and legacy
     * Forge UI payloads remain visible.
     */
    fun suppressProgressiveTransport(markdown: String): String =
        MarkdownFenceParser.parse(markdown).joinToString(separator = "") { part ->
            when (part) {
                is MarkdownFencePart.Text -> part.value
                is MarkdownFencePart.Fence -> when {
                    !part.closed -> part.raw
                    part.language == "forge-report" -> ""
                    part.language == "forge-data" && isProgressiveDataBody(part.body) -> ""
                    else -> part.raw
                }
            }
        }

    fun parse(markdown: String): List<TranscriptEnvelopePart> {
        if (markdown.isEmpty()) return emptyList()
        val result = mutableListOf<TranscriptEnvelopePart>()
        val dataBlocks = mutableListOf<TranscriptForgeDataBlock>()
        val dataRaw = mutableListOf<String>()
        var renderedUi = false
        MarkdownFenceParser.parse(markdown).forEach { part ->
            when (part) {
                is MarkdownFencePart.Text -> {
                    if (dataBlocks.isNotEmpty() && part.value.isBlank()) {
                        dataRaw += part.value
                    } else {
                        flushPendingData(result, dataBlocks, dataRaw)
                        appendMarkdown(result, part.value)
                    }
                }
                is MarkdownFencePart.Fence -> when {
                    !part.closed -> {
                        flushPendingData(result, dataBlocks, dataRaw)
                        appendMarkdown(result, part.raw)
                    }
                    part.language == "forge-data" -> parseData(part.header, part.body)?.let {
                        dataBlocks += it
                        dataRaw += part.raw
                    } ?: run {
                        flushPendingData(result, dataBlocks, dataRaw)
                        appendMarkdown(result, part.raw)
                    }
                    part.language == "forge-ui" -> parseUi(part.body)?.let { payload ->
                        result += TranscriptEnvelopePart.ForgeUi(payload, materialize(dataBlocks))
                        // A data snapshot must be immediately adjacent to one UI block.
                        dataBlocks.clear()
                        dataRaw.clear()
                        renderedUi = true
                    } ?: run {
                        flushPendingData(result, dataBlocks, dataRaw)
                        appendMarkdown(result, part.raw)
                    }
                    else -> {
                        flushPendingData(result, dataBlocks, dataRaw)
                        appendMarkdown(result, part.raw)
                    }
                }
            }
        }
        flushPendingData(result, dataBlocks, dataRaw)
        return if (renderedUi) result else listOf(TranscriptEnvelopePart.Markdown(markdown))
    }

    /**
     * Builds the same envelope from the canonical SDK representation. Raw
     * parsing is reserved for streaming or legacy messages where that additive
     * representation is not available.
     */
    fun fromCanonical(parts: List<TranscriptCanonicalPart>): List<TranscriptEnvelopePart> {
        if (parts.isEmpty()) return emptyList()
        val result = mutableListOf<TranscriptEnvelopePart>()
        val dataBlocks = mutableListOf<TranscriptForgeDataBlock>()
        val dataRaw = mutableListOf<String>()
        var renderedUi = false
        var renderedReport = false
        parts.forEach { part ->
            when (part.kind.trim().lowercase()) {
                "markdown" -> {
                    val text = part.text.orEmpty()
                    if (dataBlocks.isNotEmpty() && text.isBlank()) {
                        dataRaw += text
                    } else {
                        flushPendingData(result, dataBlocks, dataRaw)
                        appendMarkdown(result, text)
                    }
                }

                "forgedata" -> if (isProgressiveData(part)) {
                    renderedReport = true
                } else canonicalData(part)?.let { block ->
                    dataBlocks += block
                    dataRaw += part.source.orEmpty()
                } ?: run {
                    flushPendingData(result, dataBlocks, dataRaw)
                    appendMarkdown(result, part.source.orEmpty())
                }

                "forgeui" -> canonicalUi(part)?.let { payload ->
                    result += TranscriptEnvelopePart.ForgeUi(payload, materialize(dataBlocks))
                    dataBlocks.clear()
                    dataRaw.clear()
                    renderedUi = true
                } ?: run {
                    flushPendingData(result, dataBlocks, dataRaw)
                    appendMarkdown(result, part.source.orEmpty())
                }

                "forgereport" -> renderedReport = true

                else -> {
                    flushPendingData(result, dataBlocks, dataRaw)
                    appendMarkdown(result, part.source.orEmpty())
                }
            }
        }
        flushPendingData(result, dataBlocks, dataRaw)
        return if (renderedUi || renderedReport) result else listOf(
            TranscriptEnvelopePart.Markdown(parts.joinToString(separator = "") { it.text ?: it.source.orEmpty() })
        )
    }

    fun rows(value: Any?): List<Map<String, Any?>> = when (value) {
        is List<*> -> value.filterIsInstance<Map<String, Any?>>()
        is Map<*, *> -> listOf(value.entries.associate { it.key.toString() to it.value })
        else -> emptyList()
    }

    private fun parseData(header: String, body: String): TranscriptForgeDataBlock? {
        val attributes = MarkdownFenceHeader.attributes(header)
        val headerID = attributes["id"]?.trim().orEmpty()
        if (headerID.isNotEmpty()) {
            val format = attributes["format"]?.trim()?.lowercase()
            val data = if (format == "csv") JsonPrimitive(body) else runCatching { json.parseToJsonElement(body) }.getOrNull() ?: return null
            return TranscriptForgeDataBlock(id = headerID, format = format, mode = attributes["mode"], data = data)
        }
        val dataObject = runCatching { json.parseToJsonElement(body) as? JsonObject }.getOrNull()
        if (dataObject != null && dataObject["id"] != null) {
            return runCatching {
                json.decodeFromJsonElement(TranscriptForgeDataBlock.serializer(), dataObject)
            }.getOrNull()?.takeIf { !it.id.isNullOrBlank() }
        }
        return null
    }

    private fun parseUi(body: String): TranscriptForgeUiPayload? = runCatching {
        json.decodeFromString(TranscriptForgeUiPayload.serializer(), body)
    }.getOrNull()

    private fun canonicalData(part: TranscriptCanonicalPart): TranscriptForgeDataBlock? {
        val data = part.data ?: return null
        val id = data.id.trim()
        if (id.isBlank() || data.payload == null) return null
        return TranscriptForgeDataBlock(
            version = data.version ?: 1,
            id = id,
            format = data.format?.trim()?.lowercase(),
            mode = data.mode,
            data = data.payload
        )
    }

    private fun isProgressiveData(part: TranscriptCanonicalPart): Boolean {
        val data = part.data ?: return false
        return data.version == 2 || !data.reportRef.isNullOrBlank()
    }

    private fun isProgressiveDataBody(body: String): Boolean {
        val data = runCatching { json.parseToJsonElement(body) as? JsonObject }.getOrNull() ?: return false
        val version = (data["version"] as? JsonPrimitive)?.content?.toIntOrNull()
        return version == 2 || !(data["reportRef"] as? JsonPrimitive)?.content.isNullOrBlank()
    }

    private fun canonicalUi(part: TranscriptCanonicalPart): TranscriptForgeUiPayload? {
        val payload = part.payload ?: return null
        return runCatching {
            json.decodeFromJsonElement(TranscriptForgeUiPayload.serializer(), payload)
        }.getOrNull()
    }

    private fun materialize(blocks: List<TranscriptForgeDataBlock>): Map<String, TranscriptForgeDataStore> {
        val store = linkedMapOf<String, TranscriptForgeDataStore>()
        blocks.forEach { block ->
            val id = block.id?.trim().orEmpty()
            if (id.isEmpty()) return@forEach
            val next = materialize(block)
            val current = store[id]?.rows
            val rows = when (block.mode?.trim()?.lowercase()) {
                "append" -> if (current is List<*> && next is List<*>) current + next else next
                "patch" -> if (current is Map<*, *> && next is Map<*, *>) current + next else next
                else -> next
            }
            store[id] = TranscriptForgeDataStore(id, rows)
        }
        return store
    }

    private fun materialize(block: TranscriptForgeDataBlock): Any? {
        val format = block.format?.trim()?.lowercase().orEmpty().ifBlank { if (block.data is JsonPrimitive) "csv" else "json" }
        if (format != "csv") return block.data?.let(JsonUtil::elementToAny)
        val text = (block.data as? JsonPrimitive)?.content.orEmpty()
        return parseCsv(text) ?: text
    }

    internal fun materializeCanonicalPayload(format: String?, payload: JsonElement?): JsonElement? {
        payload ?: return null
        if (format?.trim()?.lowercase() != "csv") return payload
        val text = (payload as? JsonPrimitive)?.content ?: return payload
        return parseCsv(text)?.let(JsonUtil::anyToElement) ?: payload
    }

    private fun parseCsv(text: String): List<Map<String, Any?>>? {
        val records = mutableListOf<List<String>>()
        val row = mutableListOf<String>()
        val current = StringBuilder()
        var quoted = false
        var index = 0
        while (index < text.length) {
            val char = text[index]
            val next = text.getOrNull(index + 1)
            when {
                char == '\"' && quoted && next == '\"' -> { current.append(char); index += 2 }
                char == '\"' -> { quoted = !quoted; index++ }
                char == ',' && !quoted -> { row += current.toString().trim(); current.clear(); index++ }
                (char == '\n' || char == '\r') && !quoted -> {
                    row += current.toString().trim()
                    if (row.any(String::isNotBlank)) records += row.toList()
                    row.clear(); current.clear()
                    index += if (char == '\r' && next == '\n') 2 else 1
                }
                else -> { current.append(char); index++ }
            }
        }
        if (quoted) return null
        row += current.toString().trim()
        if (row.any(String::isNotBlank)) records += row
        if (records.isEmpty()) return emptyList()
        val headers = records.first()
        if (headers.isEmpty() || headers.size != headers.toSet().size) return null
        return records.drop(1).map { cells ->
            headers.mapIndexed { position, header -> header to csvValue(cells.getOrElse(position) { "" }) }.toMap()
        }
    }

    private fun csvValue(value: String): Any = when {
        value.equals("true", true) -> true
        value.equals("false", true) -> false
        value.toLongOrNull() != null -> value.toLong()
        value.toDoubleOrNull() != null -> value.toDouble()
        else -> value
    }

    private fun appendMarkdown(parts: MutableList<TranscriptEnvelopePart>, text: String) {
        if (text.isBlank()) return
        val previous = parts.lastOrNull()
        if (previous is TranscriptEnvelopePart.Markdown) parts[parts.lastIndex] = previous.copy(text = previous.text + text) else parts += TranscriptEnvelopePart.Markdown(text)
    }

    private fun flushPendingData(
        parts: MutableList<TranscriptEnvelopePart>,
        blocks: MutableList<TranscriptForgeDataBlock>,
        raw: MutableList<String>
    ) {
        raw.forEach { appendMarkdown(parts, it) }
        blocks.clear()
        raw.clear()
    }
}

File: ios/Sources/ForgeIOSRuntime/TranscriptEnvelope.swift
import Foundation

public struct TranscriptForgeDataBlock: Codable, Equatable, Sendable {
    public let version: Int?
    public let id: String?
    public let format: String?
    public let mode: String?
    public let data: JSONValue?

    public init(version: Int? = nil, id: String? = nil, format: String? = nil, mode: String? = nil, data: JSONValue? = nil) {
        self.version = version
        self.id = id
        self.format = format
        self.mode = mode
        self.data = data
    }
}

public struct TranscriptForgeUIPayload: Codable, Equatable, Sendable {
    public let version: Int?
    public let title: String?
    public let subtitle: String?
    public let blocks: [JSONValue]

    public init(version: Int? = nil, title: String? = nil, subtitle: String? = nil, blocks: [JSONValue] = []) {
        self.version = version
        self.title = title
        self.subtitle = subtitle
        self.blocks = blocks
    }
}

public struct TranscriptForgeDataStore: Equatable, Sendable {
    public let id: String
    public let rows: JSONValue

    public init(id: String, rows: JSONValue) {
        self.id = id
        self.rows = rows
    }
}

/// Platform-neutral transcript parts emitted by a conversation SDK. Forge owns
/// their data/UI pairing so native hosts do not re-parse completed messages.
public struct TranscriptCanonicalData: Equatable, Sendable {
    public let version: Int?
    public let scope: String?
    public let reportRef: String?
    public let sequence: Int?
    public let id: String
    public let format: String?
    public let mode: String?
    public let payload: JSONValue?

    public init(version: Int? = nil, scope: String? = nil, reportRef: String? = nil, sequence: Int? = nil, id: String, format: String? = nil, mode: String? = nil, payload: JSONValue? = nil) {
        self.version = version
        self.scope = scope
        self.reportRef = reportRef
        self.sequence = sequence
        self.id = id
        self.format = format
        self.mode = mode
        self.payload = payload
    }
}

public struct TranscriptCanonicalPart: Equatable, Sendable {
    public let kind: String
    public let text: String?
    public let source: String?
    public let payload: JSONValue?
    public let data: TranscriptCanonicalData?

    public init(kind: String, text: String? = nil, source: String? = nil, payload: JSONValue? = nil, data: TranscriptCanonicalData? = nil) {
        self.kind = kind
        self.text = text
        self.source = source
        self.payload = payload
        self.data = data
    }
}

public struct TranscriptCanonicalReport: Equatable, Sendable {
    public let scope: String
    public let id: String
    public let grammar: String
    public let status: String
    public let sequence: Int?
    public let resetVersion: Int
    public let source: JSONValue
    public let dataSources: [String: TranscriptCanonicalData]

    public init(
        scope: String,
        id: String,
        grammar: String,
        status: String,
        sequence: Int? = nil,
        resetVersion: Int = 0,
        source: JSONValue,
        dataSources: [String: TranscriptCanonicalData] = [:]
    ) {
        self.scope = scope
        self.id = id
        self.grammar = grammar
        self.status = status
        self.sequence = sequence
        self.resetVersion = resetVersion
        self.source = source
        self.dataSources = dataSources
    }
}

public enum TranscriptEnvelopePart: Equatable, Sendable {
    case markdown(String)
    case forgeUI(TranscriptForgeUIPayload, [String: TranscriptForgeDataStore])
}

/// Generic Forge transcript envelope decoding. Hosts decide where to place the
/// result; Forge owns fence interpretation and data-mode materialization.
public enum TranscriptEnvelope {
    /// Removes progressive report transport fences from text that accompanies
    /// an already-assembled canonical report. Ordinary markdown and legacy
    /// Forge UI payloads remain visible.
    public static func suppressProgressiveTransport(in markdown: String) -> String {
        MarkdownFenceParser.parse(markdown).map { part in
            switch part {
            case .text(let value):
                return value
            case .fence(let raw, let language, _, let body, let closed):
                guard closed else { return raw }
                if language == "forge-report" {
                    return ""
                }
                if language == "forge-data", isProgressiveDataBody(body) {
                    return ""
                }
                return raw
            }
        }.joined()
    }

    public static func parse(_ markdown: String) -> [TranscriptEnvelopePart] {
        guard !markdown.isEmpty else { return [] }
        var parts: [TranscriptEnvelopePart] = []
        var dataBlocks: [TranscriptForgeDataBlock] = []
        var dataRaw: [String] = []
        var renderedUI = false
        for part in MarkdownFenceParser.parse(markdown) {
            switch part {
            case .text(let value):
                if !dataBlocks.isEmpty, value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    dataRaw.append(value)
                } else {
                    flushPendingData(&parts, dataBlocks: &dataBlocks, raw: &dataRaw)
                    appendMarkdown(&parts, value)
                }
            case .fence(let raw, let language, let header, let body, let closed):
                guard closed else { flushPendingData(&parts, dataBlocks: &dataBlocks, raw: &dataRaw); appendMarkdown(&parts, raw); continue }
                switch language {
                case "forge-data":
                    guard let data = decodeData(header: header, body: body), let id = data.id?.trimmingCharacters(in: .whitespacesAndNewlines), !id.isEmpty else { flushPendingData(&parts, dataBlocks: &dataBlocks, raw: &dataRaw); appendMarkdown(&parts, raw); continue }
                    dataBlocks.append(data)
                    dataRaw.append(raw)
                case "forge-ui":
                    guard let payload = try? JSONDecoder().decode(TranscriptForgeUIPayload.self, from: Data(body.utf8)) else { flushPendingData(&parts, dataBlocks: &dataBlocks, raw: &dataRaw); appendMarkdown(&parts, raw); continue }
                    parts.append(.forgeUI(payload, materialize(dataBlocks)))
                    // Data fences are a snapshot for the next UI fence. A later
                    // UI block must supply its own data rather than inherit rows.
                    dataBlocks.removeAll(keepingCapacity: true)
                    dataRaw.removeAll(keepingCapacity: true)
                    renderedUI = true
                default: flushPendingData(&parts, dataBlocks: &dataBlocks, raw: &dataRaw); appendMarkdown(&parts, raw)
                }
            }
        }
        flushPendingData(&parts, dataBlocks: &dataBlocks, raw: &dataRaw)
        return renderedUI ? parts : [.markdown(markdown)]
    }

    /// Builds the same envelope from the canonical SDK representation. Raw
    /// parsing is reserved for streaming or legacy messages without that field.
    public static func fromCanonical(_ canonicalParts: [TranscriptCanonicalPart]) -> [TranscriptEnvelopePart] {
        guard !canonicalParts.isEmpty else { return [] }
        var parts: [TranscriptEnvelopePart] = []
        var dataBlocks: [TranscriptForgeDataBlock] = []
        var dataRaw: [String] = []
        var renderedUI = false
        var renderedReport = false
        for part in canonicalParts {
            switch part.kind.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
            case "markdown":
                let text = part.text ?? ""
                if !dataBlocks.isEmpty, text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    dataRaw.append(text)
                } else {
                    flushPendingData(&parts, dataBlocks: &dataBlocks, raw: &dataRaw)
                    appendMarkdown(&parts, text)
                }
            case "forgedata":
                if isProgressiveData(part) {
                    renderedReport = true
                    continue
                } else if let data = canonicalData(part) {
                    dataBlocks.append(data)
                    dataRaw.append(part.source ?? "")
                } else {
                    flushPendingData(&parts, dataBlocks: &dataBlocks, raw: &dataRaw)
                    appendMarkdown(&parts, part.source ?? "")
                }
            case "forgeui":
                if let payload = canonicalUI(part) {
                    parts.append(.forgeUI(payload, materialize(dataBlocks)))
                    dataBlocks.removeAll(keepingCapacity: true)
                    dataRaw.removeAll(keepingCapacity: true)
                    renderedUI = true
                } else {
                    flushPendingData(&parts, dataBlocks: &dataBlocks, raw: &dataRaw)
                    appendMarkdown(&parts, part.source ?? "")
                }
            case "forgereport":
                renderedReport = true
                continue
            default:
                flushPendingData(&parts, dataBlocks: &dataBlocks, raw: &dataRaw)
                appendMarkdown(&parts, part.source ?? "")
            }
        }
        flushPendingData(&parts, dataBlocks: &dataBlocks, raw: &dataRaw)
        return renderedUI || renderedReport ? parts : [.markdown(canonicalParts.map { $0.text ?? $0.source ?? "" }.joined())]
    }

    public static func rows(from value: JSONValue) -> [[String: JSONValue]] {
        switch value { case .array(let entries): return entries.compactMap(\.objectValue); case .object(let row): return [row]; default: return [] }
    }

    private static func decodeData(header: String, body: String) -> TranscriptForgeDataBlock? {
        let attributes = MarkdownFenceHeader.attributes(header)
        if let id = attributes["id"]?.trimmingCharacters(in: .whitespacesAndNewlines), !id.isEmpty {
            let format = attributes["format"]?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            let data: JSONValue
            if format == "csv" {
                data = .string(body)
            } else if let decoded = try? JSONDecoder().decode(JSONValue.self, from: Data(body.utf8)) {
                data = decoded
            } else {
                return nil
            }
            return TranscriptForgeDataBlock(version: 1, id: id, format: format, mode: attributes["mode"], data: data)
        }
        guard let block = try? JSONDecoder().decode(TranscriptForgeDataBlock.self, from: Data(body.utf8)),
              let id = block.id?.trimmingCharacters(in: .whitespacesAndNewlines),
              !id.isEmpty else { return nil }
        return TranscriptForgeDataBlock(version: block.version, id: id, format: block.format, mode: block.mode, data: block.data)
    }

    private static func canonicalData(_ part: TranscriptCanonicalPart) -> TranscriptForgeDataBlock? {
        guard let data = part.data else { return nil }
        let id = data.id.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !id.isEmpty, let payload = data.payload else { return nil }
        return TranscriptForgeDataBlock(
            version: data.version ?? 1,
            id: id,
            format: data.format?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased(),
            mode: data.mode,
            data: payload
        )
    }

    private static func isProgressiveData(_ part: TranscriptCanonicalPart) -> Bool {
        guard let data = part.data else { return false }
        return data.version == 2 || !(data.reportRef?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true)
    }

    private static func isProgressiveDataBody(_ body: String) -> Bool {
        guard let object = try? JSONSerialization.jsonObject(with: Data(body.utf8)) as? [String: Any] else {
            return false
        }
        if let version = object["version"] as? NSNumber, version.intValue == 2 {
            return true
        }
        guard let reportRef = object["reportRef"] as? String else {
            return false
        }
        return !reportRef.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    private static func canonicalUI(_ part: TranscriptCanonicalPart) -> TranscriptForgeUIPayload? {
        guard let payload = part.payload,
              let data = try? JSONEncoder().encode(payload) else { return nil }
        return try? JSONDecoder().decode(TranscriptForgeUIPayload.self, from: data)
    }

    private static func materialize(_ blocks: [TranscriptForgeDataBlock]) -> [String: TranscriptForgeDataStore] {
        var store: [String: TranscriptForgeDataStore] = [:]
        for block in blocks {
            guard let id = block.id?.trimmingCharacters(in: .whitespacesAndNewlines), !id.isEmpty else { continue }
            let next = materialize(block)
            let mode = block.mode?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() ?? "replace"
            let rows: JSONValue
            switch (mode, store[id]?.rows) {
            case ("append", .some(.array(let existing))):
                if case .array(let incoming) = next { rows = .array(existing + incoming) } else { rows = next }
            case ("patch", .some(.object(let existing))):
                if case .object(let incoming) = next { rows = .object(existing.merging(incoming) { _, replacement in replacement }) } else { rows = next }
            default: rows = next
            }
            store[id] = TranscriptForgeDataStore(id: id, rows: rows)
        }
        return store
    }

    private static func materialize(_ block: TranscriptForgeDataBlock) -> JSONValue {
        let format = block.format?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let inferredFormat = format?.isEmpty == false ? format! : (block.data?.stringValue == nil ? "json" : "csv")
        guard inferredFormat == "csv", let text = block.data?.stringValue else {
            return block.data ?? .null
        }
        guard let rows = parseCSVRows(text) else { return .string(text) }
        return .array(rows.map(JSONValue.object))
    }

    static func materializeCanonicalPayload(format: String?, payload: JSONValue?) -> JSONValue? {
        guard let payload else { return nil }
        let normalizedFormat = format?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard normalizedFormat == "csv", let text = payload.stringValue else { return payload }
        guard let rows = parseCSVRows(text) else { return payload }
        return .array(rows.map(JSONValue.object))
    }

    private static func parseCSVRows(_ text: String) -> [[String: JSONValue]]? {
        var records: [[String]] = []
        var row: [String] = []
        var current = ""
        var quoted = false
        var cursor = text.startIndex
        while cursor < text.endIndex {
            let character = text[cursor]
            let next = text.index(after: cursor)
            if character == "\"", quoted, next < text.endIndex, text[next] == "\"" {
                current.append("\"")
                cursor = text.index(after: next)
            } else if character == "\"" {
                quoted.toggle()
                cursor = next
            } else if character == ",", !quoted {
                row.append(current.trimmingCharacters(in: .whitespaces))
                current = ""
                cursor = next
            } else if (character == "\n" || character == "\r"), !quoted {
                row.append(current.trimmingCharacters(in: .whitespaces))
                if row.contains(where: { !$0.isEmpty }) { records.append(row) }
                row = []
                current = ""
                cursor = character == "\r" && next < text.endIndex && text[next] == "\n" ? text.index(after: next) : next
            } else {
                current.append(character)
                cursor = next
            }
        }
        guard !quoted else { return nil }
        row.append(current.trimmingCharacters(in: .whitespaces))
        if row.contains(where: { !$0.isEmpty }) { records.append(row) }
        guard let headers = records.first, !headers.isEmpty else { return [] }
        guard Set(headers).count == headers.count else { return nil }
        return records.dropFirst().map { cells in
            Dictionary(uniqueKeysWithValues: headers.enumerated().map { index, header in
                (header, csvValue(cells.indices.contains(index) ? cells[index] : ""))
            })
        }
    }

    private static func csvValue(_ raw: String) -> JSONValue {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if text.caseInsensitiveCompare("true") == .orderedSame { return .bool(true) }
        if text.caseInsensitiveCompare("false") == .orderedSame { return .bool(false) }
        if let value = Double(text) { return .number(value) }
        return .string(text)
    }

    private static func appendMarkdown(_ parts: inout [TranscriptEnvelopePart], _ text: String) {
        guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }
        if case .markdown(let existing)? = parts.last { parts[parts.count - 1] = .markdown(existing + text) } else { parts.append(.markdown(text)) }
    }

    private static func flushPendingData(
        _ parts: inout [TranscriptEnvelopePart],
        dataBlocks: inout [TranscriptForgeDataBlock],
        raw: inout [String]
    ) {
        raw.forEach { appendMarkdown(&parts, $0) }
        dataBlocks.removeAll(keepingCapacity: true)
        raw.removeAll(keepingCapacity: true)
    }
}
